### PR TITLE
Sos 218 update spending pie chart to use new subcategories

### PR DESCRIPTION
**Pull request by SeanTConrad**

### CODE REVIEW

File: client/src/components/types.ts [hunk 1]
```diff
@@ -68,6 +68,8 @@ export interface TransactionType {
   created_at: string;
   updated_at: string;
   merchant: string;
+  sos_category: string;
```
Comment (aylusltd) [inline client/src/components/types.ts:71]: should make this an enum

File: client/src/components/WidgetDescription.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/*
+    Widget explination text for report dashboard.
```
Comment (aylusltd) [inline client/src/components/WidgetDescription.tsx:2]: explanation

File: client/src/components/WidgetDescription.test.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import '../matchMedia.mock.js'; // Must be imported before the tested file
+import { render, screen } from '@testing-library/react';
+import '@testing-library/jest-dom'
+
+import WidgetDescription from './WidgetDescription';
+import ListItemIcon from '@mui/material/ListItemIcon';
+
+
+it('renders WidgetDescription', () => {
+    render(<WidgetDescription>This is just some text.</WidgetDescription>);
```
Comment (aylusltd) [inline client/src/components/WidgetDescription.test.tsx:10]: these should be different tests

File: client/src/components/SpendingBreakdown.tsx [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+              date > oneYearAgo &&
+              tx.category !== 'Payment' &&
+              tx.category !== 'Transfer' &&
+              tx.category !== 'Interest'
+            );
+          }),
+        [transactions]
+    );
+
+    let arr = transactions.map(t=>Date.parse(t.date))
```
Comment (aylusltd) [inline client/src/components/SpendingBreakdown.tsx:64]: Do we still need this?

File: client/src/components/WidgetDescription.test.tsx (hunk 3 above)
Comment (aylusltd) [inline client/src/components/WidgetDescription.test.tsx:10]: these should be different tests

**Review by aylusltd — CHANGES_REQUESTED** [review]

File: client/src/components/WidgetDescription.test.tsx (hunk 3 above)
Comment (aylusltd) [inline client/src/components/WidgetDescription.test.tsx:10]: These should be separate tests

**Review by aylusltd — CHANGES_REQUESTED** [review]

File: client/src/components/types.ts (hunk 1 above)
Comment (SeanTConrad) [reply client/src/components/types.ts:71]: Just tried, but even with bogus values, it just accepts what the API sends back:
image

File: client/src/components/SpendingBreakdown.tsx [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    for (let i = 0; i < expensePercents.length; i++) {
+
+        // Highcharts aligns inner and out pies by data values, so switch
+        // color base of expenses based on underlying category
+        const cumeExpenses = expensePercents.reduce((p,c,index)=>{ return index<=i?p+c:p });
+
+        // Determine which category we are in with matched value sums
+        let categoryGroupIndex = 0;
+        for(let j = 0; j < categoryPercents.length; j++){
+            const cumCategories = categoryPercents.reduce((p,c,index)=>{ return index<=j?p+c:p });
```
Comment (aylusltd) [inline client/src/components/SpendingBreakdown.tsx:204]: I have to imagine there's a better name for this...

File: client/src/components/SpendingBreakdown.tsx [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+    let expenseFormattedLabels:string[] = [];
+    for (var i = 0; i < expensePercents.length; i++) {
+        expenseFormattedLabels.push("<b>"+expenseLabels[i]+"</b>: " + 
+            numberToDollarStr(Math.round(expenseAmounts[i]),0) + " (" +
+            (Math.round(expensePercents[i] * 10)/10) + "%)"
+        )
+    }
+
+  const options: Highcharts.Options = {
```
Comment (aylusltd) [inline client/src/components/SpendingBreakdown.tsx:243]: should this go in a theme?

**Review by aylusltd — CHANGES_REQUESTED** [review]
Minor stuff

File: client/src/components/WidgetDescription.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+/*
+    Widget explanation text for report dashboard.
+*/
+
+import Typography from '@mui/material/Typography';
+
+interface IWidgetDescriptionProps {
+    children: string | JSX.Element | JSX.Element[] | (string | JSX.Element)[];
```
[the remark this answers is not in the document]
Comment (SeanTConrad) [reply client/src/components/WidgetDescription.tsx:8]: The union is an array of strings or JSX Elements.

File: client/src/components/SpendingBreakdown.tsx [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        return ttmTransactions.reduce((total, tx) => tx.amount + total,0)
+    }, [ttmTransactions]);
+
+    // Uncomment to QA the totals with DB
+    // console.log("totalExpenses", totalExpenses);
+    // const qaHostingCosts = transactions.reduce((total, obj) => 
+    //     obj.sos_subcategory==='Hosting Costs'?obj.amount + total:total,0
+    // );
+
+    const categoriesObject = useMemo((): CategorySummary => {
```
[the remark this answers is not in the document]
Comment (SeanTConrad) [reply client/src/components/SpendingBreakdown.tsx:89]: I check this against a database query, so I don't have a value to match against in the unit test. It's really difficult to get values out of a highchart as well.

File: client/src/components/SpendingBreakdown.tsx [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+            }
+        )
+        expenseList.map((obj2)=>{
+            expenseLabels.push(obj2[0]);
+            expensePercents.push(obj2[1].percent);
+            expenseAmounts.push(obj2[1].amount);
+        });
+    });
+
+    const categoryColorLevel = 800;
```
[the remark this answers is not in the document]
Comment (SeanTConrad) [reply client/src/components/SpendingBreakdown.tsx:156]: Chart colors aren't in the themes yet. This was just a quick way to get color gradients for the subcategories.

File: client/src/components/SpendingBreakdown.tsx [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+    let expensePercentsWColor:formattedHighChartSlice[] = [];
+
+    let colorOffset = 1;
+    let prevCategoryGroupIndex = 0;
+    for (let i = 0; i < expensePercents.length; i++) {
+
+        // Highcharts aligns inner and out pies by data values, so switch
+        // color base of expenses based on underlying category
+        const cumeExpenses = expensePercents.reduce((p,c,index)=>{ return index<=i?p+c:p });
```
[the remark this answers is not in the document]
Comment (SeanTConrad) [reply client/src/components/SpendingBreakdown.tsx:199]: I think I fixed it. I am lost without a red squiggle.

File: client/src/components/SpendingBreakdown.tsx (hunk 5 above)
Comment (SeanTConrad) [reply client/src/components/SpendingBreakdown.tsx:204]: Whoops! fixed

File: client/src/components/SpendingBreakdown.tsx (hunk 6 above)
Comment (SeanTConrad) [reply client/src/components/SpendingBreakdown.tsx:243]: same as above. Someday I'll add a theme "base red", but it's a low priority. Current way helps us keep it close to ADA compliance too.

File: client/src/components/WidgetDescription.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (aylusltd) [reply client/src/components/WidgetDescription.tsx:8]: right, which should accept an array of JSX Elements as well

File: client/src/components/SpendingBreakdown.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (aylusltd) [reply client/src/components/SpendingBreakdown.tsx:199]: hasn't changed here. would also prefer a different var name if not too much trouble...

File: client/src/components/WidgetDescription.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (SeanTConrad) [reply client/src/components/WidgetDescription.tsx]: Ah, fixed.

**Pull request merged**